include/osquery/expected.h

```diff
@@ -0,0 +1,164 @@
+//
+//  expected.hpp
+//  gmock
+//
+//  Created by Max Kareta on 5/15/18.
+//
+
+#pragma once
+
+#include <include/osquery/error.h>
+
+/** Utility class that should be used in function that return
+ * either error or value. Expected enforce developer to test for success and
+ * check error if any.
+ *
+ * ExpectedValue:
+ * ExpectedValue<std::string> function() {
+ *   if (test) {
+ *    return "ok";
+ *   } else {
+ *    return std::shared_prt(new Error(error_domain, error_code));
+ *   }
+ * }
+ *
+ * Expected:
+ * Expected<PlatformProcess> function() {
+ *   if (test) {
+ *    return std::unique_prt(new PlatformProcess(pid));
+ *   } else {
+ *    return std::shared_prt(new Error(error_domain, error_code));
+ *   }
+ * }
+ *
+ * auto result = function();
+ * if (result) {
+ *   ...use *result
+ * } else {
+ *   auto error = result->getError();
+ * }
+ */
+
+namespace osquery {
+
+class ExpectedBase {
+ public:
+  ExpectedBase(Error* error) = delete;
+  ExpectedBase(Error error) = delete;
+  ExpectedBase() : error_(nullptr), hasError_(false) {}
+  ExpectedBase(std::shared_ptr<Error> error)
+      : error_(std::move(error)), hasError_(true) {}
+  ExpectedBase(std::unique_ptr<Error> error)
+      : error_(std::move(error)), hasError_(true) {}
+
+  ~ExpectedBase() {
+    assert(errorChecked_ && (!hasError_ || errorUsed_));
+  }
+
+  std::shared_ptr<Error> getError() {
+    errorUsed_ = true;
+    errorChecked_ = true;
+    return error_;
+  }
+
+  explicit operator bool() {
+    errorChecked_ = true;
+    return !hasError_;
+  }
+
+ protected:
+  std::shared_ptr<Error> error_;
```

Comment:
Is here any reason to share owership of Error object with someone? Is not it better to use `std::unique_ptr` here instead?

Reply:
Unique pointer will lead to take_error semantic, moreover it will add limitaion for source of error to be onle unique pointer.  Error handling should be always slow path so performance diffirence is completely neglegable

Reply:
In most cases this is a semantic difference between shared and unique. First one is for sharing and second one is just for owning. I'm just wonder do we need to share ownership with someone here?

Reply:
Unique prt will work for methods that are creating new object as result, but will fail for cases when method should return pointer to existing class, which will very likely will be already shared object.

Reply:
Hmm, it makes sense. May it is a good reason to make the whole class template?
```c++
template<
  typename ValueType,
  typename ErrorType
>
class Expected {
  ...
};
```

And usage would be look like:
```c++
class TypeError;
class Widget;

Expected<Widget, TypeError> widgetFromString(const std::string& str) {
  ...
  if (smt_wrong) {
    return TypeError("Wrong widget text representation");
  }
  return Widget(...);
}

What do you think?